pkg/logql/log/parser_test.go

```diff
@@ -722,6 +721,163 @@ func Test_logfmtParser_Parse(t *testing.T) {
 	}
 }
 
+func TestLogfmtExpressionParser(t *testing.T) {
+	testLine := []byte(`app=foo level=error spaces="value with ÜFT8👌" ts=2021-02-12T19:18:10.037940878Z`)
+
+	tests := []struct {
+		name        string
+		line        []byte
+		expressions []LogfmtExpression
+		lbs         labels.Labels
+		want        labels.Labels
+	}{
+		{
+			"single field",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("app", "app"),
+			},
+			labels.Labels{},
+			labels.Labels{
+				{Name: "app", Value: "foo"},
+			},
+		},
+		{
+			"empty line",
+			[]byte("{}"),
+			[]LogfmtExpression{},
+			labels.Labels{},
+			labels.Labels{},
+		},
+		{
+			"multiple fields",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("app", "app"),
+				NewLogfmtExpr("level", "level"),
+				NewLogfmtExpr("ts", "ts"),
+			},
+			labels.Labels{},
+			labels.Labels{
+				{Name: "app", Value: "foo"},
+				{Name: "level", Value: "error"},
+				{Name: "ts", Value: "2021-02-12T19:18:10.037940878Z"},
+			},
+		},
+		{
+			"label renaming",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("test", "level"),
+			},
+			labels.Labels{},
+			labels.Labels{
+				{Name: "test", Value: "error"},
+			},
+		},
+		{
+			"multiple fields with label renaming",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("app", "app"),
+				NewLogfmtExpr("lvl", "level"),
+				NewLogfmtExpr("timestamp", "ts"),
+			},
+			labels.Labels{},
+			labels.Labels{
+				{Name: "app", Value: "foo"},
+				{Name: "lvl", Value: "error"},
+				{Name: "timestamp", Value: "2021-02-12T19:18:10.037940878Z"},
+			},
+		},
+		{
+			"value with spaces and ÜFT8👌",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("spaces", "spaces"),
+			},
+			labels.Labels{},
+			labels.Labels{
+				{Name: "spaces", Value: "value with ÜFT8👌"},
+			},
+		},
+		{
+			"expression matching nothing",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("nope", "nope"),
+			},
+			labels.Labels{},
+			labels.Labels{},
+		},
+		{
+			"double property logfmt",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("app", "app"),
+			},
+			labels.Labels{
+				{Name: "ap", Value: "bar"},
+			},
+			labels.Labels{
+				{Name: "ap", Value: "bar"},
+				{Name: "app", Value: "foo"},
+			},
+		},
+		{
+			"label override",
+			testLine,
+			[]LogfmtExpression{
+				NewLogfmtExpr("app", "app"),
+			},
+			labels.Labels{
+				{Name: "app", Value: "bar"},
+			},
+			labels.Labels{
+				{Name: "app", Value: "bar"},
+				{Name: "app_extracted", Value: "foo"},
```

Comment:
Is it expected that the `app_extracted` label is returned, even though it has not been requested in the expressions?

Reply:
The suffix `_extracted` is only added if the label we extract from is already in the log stream's labels.
I followed the footsteps of the same feature in `json`. ([here's the discussion](https://github.com/grafana/loki/pull/3280#discussion_r569746897))

Reply:
I added another test case to illustrate my concern:

```go
		{
			"label override 2",
			testLine,
			[]LogfmtExpression{
				NewLogfmtExpr("lvl", "level"),
			},
			labels.Labels{
				{Name: "app", Value: "bar"},
			},
			labels.Labels{
				{Name: "app", Value: "bar"},
				{Name: "lvl", Value: "error"},
			},
		},
```

I would not expect `app_extracted` to be extracted, but it does:

```
=== RUN   TestLogfmtExpressionParser/label_override_2
    parser_test.go:870: 
        	Error Trace:	/home/christian/sandbox/grafana/loki/pkg/logql/log/parser_test.go:870
        	Error:      	Not equal: 
        	            	expected: labels.Labels{labels.Label{Name:"app", Value:"bar"}, labels.Label{Name:"lvl", Value:"error"}}
        	            	actual  : labels.Labels{labels.Label{Name:"app", Value:"bar"}, labels.Label{Name:"app_extracted", Value:"foo"}, labels.Label{Name:"lvl", Value:"error"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(labels.Labels) (len=2) {
        	            	+(labels.Labels) (len=3) {
        	            	  (labels.Label) {
        	            	@@ -4,2 +4,6 @@
        	            	   Value: (string) (len=3) "bar"
        	            	+ },
        	            	+ (labels.Label) {
        	            	+  Name: (string) (len=13) "app_extracted",
        	            	+  Value: (string) (len=3) "foo"
        	            	  },
        	Test:       	TestLogfmtExpressionParser/label_override_2
--- FAIL: TestLogfmtExpressionParser (0.00s)
```

Reply:
I see.. I'll take a closer look at the code and figure out why it behaved like this

Reply:
@chaudum maybe I'm misunderstanding this.. but why wouldn't `app_extracted` be extracted?

Reply:
If the original labels contains `app`, then it would only be extracted if you use `| logfmt` or `| logfmt app`, but not if you explicitly extract other labels, such as `| logftm level=lvl`

Reply:
I just tested this with `| json` and it has the same behaviour as I described.

Reply:
@chaudum I managed to correct this issue. I still have one question though:
in this use case that you've added above:
```
{
			"label override 2",
			testLine,
			[]LogfmtExpression{
				NewLogfmtExpr("lvl", "level"),
			},
			labels.Labels{
				{Name: "app", Value: "bar"},
			},
			labels.Labels{
				{Name: "app", Value: "bar"},
				{Name: "lvl", Value: "error"},
			},
		},
```
based on your comment above, I guess you meant 
```
                        labels.Labels{
				{Name: "app", Value: "foo"},
				{Name: "lvl", Value: "error"},
			},
```
 instead of `{Name: "app", Value: "bar"},` ? (if no then I am still confused on how this particular case of extraction works, what's the logic behind not extracting `app=foo` from the log line?)

Reply:
Sorry for the delay in the response.

* When using `| logfmt`, *all* labels of the log line should be extracted. If a label is already present in the stream, then the label from the log line gets the `_extracted` suffix. That is current behaviour.
* When using `| logfmt name`, then only the label `name` should be extracted. If the label is already present in the stream, then the label from the log lines becomes `name_extracted`. No other labels must be extracted.

In the case of this test case, the line is
```
app=foo level=error spaces="value with ÜFT8👌" ts=2021-02-12T19:18:10.037940878Z

```
and the stream already has the label pair `app=bar`.

Therefore, `app` must not get extracted again, only `level`, which gets renamed to `lvl`, resulting in the expected response:

```go
			labels.Labels{
				{Name: "app", Value: "bar"},
				{Name: "lvl", Value: "error"},
			},
```